src/webgpu/shader/validation/expression/call/builtin/length.spec.ts

```diff
@@ -0,0 +1,178 @@
+const builtin = 'length';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import {
+  ScalarType,
+  TypeF16,
+  TypeF32,
+  elementType,
+  kAllFloatScalars,
+  kAllFloatVector2,
+  kAllFloatVector3,
+  kAllFloatVector4,
+  kAllIntegerScalarsAndVectors,
+} from '../../../../../util/conversion.js';
+import { isRepresentable } from '../../../../../util/floating_point.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+/** @returns true if the sum of the squares of each element in @p vec is representable by @p type */
+function isSquareSumRepresentable(vec: number[], type: ScalarType) {
+  const squareSum = vec.reduce((prev, curr) => prev + curr * curr, 0);
+  return isRepresentable(squareSum, type);
+}
+
+g.test('scalar')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() with
+the input scalar value always compiles without error
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', kAllFloatScalars)
+      .filter(u => stageSupportsType(u.stage, u.type))
+      .expand('value', u => fullRangeForType(u.type))
+  )
+  .beforeAllSubcases(t => {
+    if (elementType(t.params.type) === TypeF16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    // We only validate with numbers known to be representable by the type
+    const expectedResult = true;
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [t.params.type.create(t.params.value)],
+      t.params.stage
+    );
+  });
+
+g.test('vec2')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() with a vec2 compiles with valid values
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', kAllFloatVector2)
+      .filter(u => stageSupportsType(u.stage, u.type))
+      .expand('x', u => fullRangeForType(u.type, 5))
+      .expand('y', u => fullRangeForType(u.type, 5))
+      .filter(u => isSquareSumRepresentable([u.x, u.y], elementType(u.type)))
```

Comment:
Instead of filtering, check the expected overflow

expectedResult = isSquareSumRepresentable(. ...);

Reply:
That's depending on an implementation detail. Internally the built-in could be evaluated with a higher precision than the data type.

Reply:
Kept the filter for the reason stated above, but also included negative tests where `isRepresentable(intermediate) == isRepresentable(result)`